File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";
import { capitalCase } from "change-case";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

interface SummaryResult {
  ingredients: Record<string, number>;
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: any[] = [];

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  let newName: string = "";

  for (const char of recipeName) {
    // First condition: Replace '-' and '_' with ' '
    if (char === '-' || char === '_')  {
      newName += ' ';
      continue;
    }

    let isLetter = (char.toLowerCase() >= 'a' && char.toLowerCase() <= 'z');
    let isSpace = (char === ' ');
    // If char is a letter or number, append to newName
    if (isLetter || isSpace) newName += char;
  }
  
  // Use change-case library to format the name to capital case
  // Using .toLowerCase() to prevent camelCase boundary
  newName = capitalCase(newName.toLowerCase());

  // Return null if the newName is empty, otherwise return the formatted name
  return newName.length === 0 ? null : newName;
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req: Request, res: Response) => {
  const entry = req.body;

  // Type must be either 'recipe' or 'ingredient'
  if (entry.type !== "recipe" && entry.type !== "ingredient") {
    return res.status(400).send();
  }

  // Each entry must have a unique name
  const nameExists = cookbook.some((existingEntry) => existingEntry.name === entry.name);
  if (nameExists) {
    return res.status(400).send();
  }

  if (entry.type === "ingredient") {
    // cookTime must be a number and >= 0
    if (typeof entry.cookTime !== "number" || entry.cookTime < 0) {
      return res.status(400).send();
    }
  }

  if (entry.type === "recipe") {
    // Check for duplicate names in requiredItems
    const seenNames: string[] = [];
    
    for (const item of entry.requiredItems) {
      if (seenNames.includes(item.name)) {
        return res.status(400).send();
      }
      seenNames.push(item.name);
    }
  }

  cookbook.push(entry);
  return res.status(200).send();
});

// Recursive helper function to get sub-recipes and sub-ingredients
const getSummary = (name: string, quantity: number, cookbook: any[]): SummaryResult | null => {
  const item = cookbook.find(e => e.name === name);
  if (!item) return null;

  // Ingredient
  if (item.type === "ingredient") {
    const ingredients: Record<string, number> = {};
    ingredients[name] = quantity;
    return { 
      ingredients, 
      cookTime: item.cookTime * quantity 
    };
  }

  // If recipe, then recurse
  if (item.type === "recipe") {
    // Store all ingredients in an object with keys as strings and values as numbers
    const totalIngredients: Record<string, number> = {};
    let totalCookTime = 0;

    for (const required of item.requiredItems) {
      const subResult = getSummary(required.name, required.quantity * quantity, cookbook);
      
      // If any sub-item is missing, return null
      if (!subResult) return null;

      // Recursively add cook time
      totalCookTime += subResult.cookTime;

      // Merge sub-ingredients into totalIngredients
      for (const [ingredientName, ingredientQuantity] of Object.entries(subResult.ingredients)) {
        // If we already have this ingredient, add to it. Otherwise start at 0.
        const currentQty = totalIngredients[ingredientName] || 0;
        totalIngredients[ingredientName] = currentQty + ingredientQuantity;
      }
    }
    
    return { ingredients: totalIngredients, cookTime: totalCookTime };
  }
  
  return null;
};

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req: Request, res: Response) => {
  const recipeName = req.query.name as string;

  // 1. Find the root item
  const rootItem = cookbook.find(existingNames => existingNames.name === recipeName);
  if (!rootItem || rootItem.type !== "recipe") {
    return res.status(400).send();
  }

  // 2. Call the helper function
  const result = getSummary(recipeName, 1, cookbook);

  if (!result) {
    return res.status(400).send();
  }

  // { Egg: 3 } turns into { name: "Egg", quantity: 3 }
  const formattedIngredients = Object.entries(result.ingredients).map(([name, quantity]) => ({
    name,
    quantity
  }));

  return res.status(200).json({
    name: recipeName,
    cookTime: result.cookTime,
    ingredients: formattedIngredients
  });
});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = []

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	# First condition. '-' and '_' become ' '
	recipeName = recipeName.replace('-', ' ').replace('_', ' ')

	newName = ""

	# Loop for valid characters
	for char in recipeName:
			if char.isalpha() or char.isspace():
				newName += char

	# Put words into an array
	words = newName.split()

	# Remove whitespace and capitalize
	words = [word.strip().capitalize() for word in words]

	# Join them back together with a single space
	newName = " ".join(words)

	# 4. Return None if empty, else the name
	return newName if len(newName) > 0 else None


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.get_json()

	# Type must be either 'recipe' or 'ingredient'
	type = data.get('type', '')
	if type != "recipe" and type != "ingredient":
		return "", 400

	# Each entry must have a unique name
	name = data.get('name', '')
	nameExists = any(entry.get('name') == name for entry in cookbook)
	if nameExists:
		return "", 400

	if type == "ingredient":
		# cookTime must be a number and >= 0
		cookTime = data.get('cookTime', '')
		if not isinstance(cookTime, int) or cookTime < 0:
			return "", 400

	if type == "recipe":
		# Check for duplicate names in requiredItems
		seenNames = []
		
		# Loop to check names
		for item in data.get('requiredItems', []):
			if item.get('name') in seenNames:
				return "", 400
			seenNames.append(item["name"])

	cookbook.append(data)
	return "", 200

# Recursive helper function to get sub-recipes and sub-ingredients
def getInfo(name, quantity):
	item = next((entry for entry in cookbook if entry.get('name') == name), None)
	if not item:
		return None	

	# Ingredient
	if item["type"] == "ingredient":
		ingredients = {}
		ingredients[name] = quantity
		itemCookTime = item["cookTime"] * quantity
		return { 
			"ingredients": ingredients,
			"cookTime": itemCookTime
		}

	# If recipe, then recurse
	if item["type"] == "recipe":
		# Store all ingredients in an object with keys as strings and values as numbers
		totalIngredients = {}
		totalCookTime = 0

		for required in item["requiredItems"]:
			subResult = getInfo(required["name"], required["quantity"] * quantity)
			
			# If any sub-item is missing, return None
			if not subResult:
				return None

			# Recursively add cook time
			totalCookTime += subResult["cookTime"]

			# Merge sub-ingredients into totalIngredients
			for ingredientName, ingredientQuantity in subResult["ingredients"].items():
				# If we already have this ingredient, add to it. Otherwise start at 0.
				currentQty = totalIngredients.get(ingredientName, 0)
				totalIngredients[ingredientName] = currentQty + ingredientQuantity

		return { 
			"ingredients": totalIngredients, 
			"cookTime": totalCookTime
		}
  
	return None

# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	recipeName = request.args.get('name')

	rootItem = next((entry for entry in cookbook if entry.get('name') == recipeName), None)
	if not rootItem or rootItem["type"] != "recipe":
		return "", 400

	# Use recursive helper function
	result = getInfo(recipeName, 1)

	if not result:
		return "", 400

	# { Egg: 3 } turns into { name: "Egg", quantity: 3 }
	formattedIngredients = [
		{
		"name": name,
		"quantity": quantity
		} 
		for name, quantity in result["ingredients"].items()
	]

	return jsonify({
		"name": recipeName,
		"cookTime": result["cookTime"],
		"ingredients": formattedIngredients
	}), 200
	

# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
